### PR TITLE
Set cluster connect timeout to infinite by default

### PR DESCRIPTION
**Pull request by mdumandag**
After an internal discussion, we have decided to set the cluster
connect timeout to infinite by default. Also, the multiplier
for the sleep time between consecutive connect attempts is changed
to `1.05` from `1.0`. The new sleep times between connection
attempts can be seen in the following graph. https://www.desmos.com/calculator/hl7rnbk4gf
(x-axis: attempt number, y-axis: sleep time in seconds).

The default value for the cluster connect timeout is set to `-1`. For this value,
the client will not stop trying to connect the cluster. If the failover client is used,
the default cluster connection timeout will be 2 minutes. If the timeout is set to
any other value, both the failover and non-failover client will use it as it is.

Also, updated some logging statements.

Note that, this is a breaking change for the users who rely on the client to stop
trying to connect cluster after some time, but we decided

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java [hunk 1]
```diff
@@ -281,13 +282,20 @@ protected NioNetworking initNetworking() {
...
-                expoRetryConfig.getMaxBackoffMillis(),
-                expoRetryConfig.getMultiplier(),
-                expoRetryConfig.getClusterConnectTimeoutMillis(),
-                expoRetryConfig.getJitter(), logger);
+        ConnectionRetryConfig retryConfig = clientConfig
+                .getConnectionStrategyConfig()
+                .getConnectionRetryConfig();
+
+        long clusterConnectTimeout = retryConfig.getClusterConnectTimeoutMillis();
+        if (failoverConfigProvided && !ConnectionRetryConfigAccessor.isClusterConnectTimeoutConfigured(retryConfig)) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:290]: I used that accessor approach to respect the configuration when the user sets an infinite timeout in a scenario like below.
```java
        ClientFailoverConfig failoverConfig = new ClientFailoverConfig();
        ClientConfig config = new ClientConfig();
        config.getConnectionStrategyConfig().getConnectionRetryConfig().setClusterConnectTimeoutMillis(Long.MAX_VALUE);
        failoverConfig.addClientConfig(config);
        HazelcastClient.newHazelcastFailoverClient(failoverConfig);
```
If we think that this scenario is unlikely or does not make sense, I can ditch that accessor and check whether or not cluster connect timeout equals to the default value for it.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   XmlClientConfigBuilderTest>AbstractClientConfigBuilderTest.testExponentialConnectionRetryConfig_defaults:300 expected:<120000> but was:<9223372036854775807>
[ERROR]   YamlClientConfigBuilderTest>AbstractClientConfigBuilderTest.testExponentialConnectionRetryConfig_defaults:300 expected:<120000> but was:<9223372036854775807>
[INFO] 
[ERROR] Tests run: 3682, Failures: 2, Errors: 0, Skipped: 18
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

**Review by ihsandemir — APPROVED** [review]

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed.
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ReplicatedMapMBeanTest.testAttributesAndOperations:113
[INFO] 
[ERROR] Tests run: 35453, Failures: 1, Errors: 0, Skipped: 979
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

File: hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java (hunk 1 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java:290]: Yes, it looks like a wrong configuration. We may fail fast in this case. But what happens if user sets Int.MAX for example. This probably a wrong setting also.So I am not sure about failfast behaviour. Lets leave it aside.  

How you considered making the default value -1 ? We use that approach in a couple of places.
Documentation will make it clear that, in normal setup, we default to not closing the client. 
In failover we use 120 seconds. 

That way you can track if user set it or not without accessor.

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java]: Okay, changed the default value to -1 and added documentation to the getter and setter methods. Can you take another look?

**Review by sancar — APPROVED** [review]

**Comment by mdumandag** [comment]
Thanks for the reviews, I am merging this

**Pull request merged**